tools/inspect/Inspect.cpp

```diff
@@ -287,76 +387,82 @@ int run(int argc, ACE_TCHAR* argv[])
-                                       TOPIC_QOS_DEFAULT,
-                                       0,
-                                       DEFAULT_STATUS_MASK);
-
-      if (!topic) {
-        if (log_level >= LogLevel::Error) {
-          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: main(): create_topic failed!\n"));
-        }
-        return 1;
-      }
+    if (!check_rc(ws->attach_condition(gc), "attach guard condition")) {
+      return 1;
+    }
+    if (!check_rc(ws->attach_condition(dr_rc), "attach read condition")) {
+      return 1;
+    }
 
-      RcHandle<RecorderListenerImpl> recorder_listener = make_rch<RecorderListenerImpl>(gc);
-
-      DDS::SubscriberQos sub_qos;
-      participant->get_default_subscriber_qos(sub_qos);
-
-      DDS::DataReaderQos dr_qos = service->initial_DataReaderQos();
-      dr_qos.representation.value.length(2);
-      dr_qos.representation.value[0] = DDS::XCDR_DATA_REPRESENTATION;
-      dr_qos.representation.value[1] = DDS::XCDR2_DATA_REPRESENTATION;
-      dr_qos.reliability.kind = DDS::RELIABLE_RELIABILITY_QOS;
-
-      // Create Recorder
-      Recorder_var recorder =
-        service->create_recorder(participant,
-                                 topic.in(),
-                                 sub_qos,
-                                 dr_qos,
-                                 recorder_listener);
-      if (!recorder.in()) {
-        if (log_level >= LogLevel::Error) {
-          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: main(): create_recorder failed!\n"));
-        }
-        return 1;
+    const DDS::Duration_t infinite = {DDS::DURATION_INFINITE_SEC, DDS::DURATION_INFINITE_NSEC};
+    const bool use_timeout = num_seconds > 0;
+    const SystemTimePoint end = SystemTimePoint::now() + TimeDuration(num_seconds);
+    while (true) {
+      DDS::ConditionSeq active;
+      const DDS::Duration_t timeout = use_timeout ? (end - SystemTimePoint::now()).to_dds_duration() : infinite;
+      if (!check_rc(ws->wait(active, timeout), "wait on publication BIT")) {
+        break;
       }
-
-      DDS::Duration_t timeout = { DDS::DURATION_INFINITE_SEC, DDS::DURATION_INFINITE_NSEC };
-      if (num_seconds > 0) {
-        timeout.sec = static_cast<DDS::Int32>(num_seconds);
-        timeout.nanosec = 0;
+      if (gc->get_trigger_value() || (use_timeout && SystemTimePoint::now() >= end)) {
+        ACE_DEBUG((LM_DEBUG, "Reached stop condition\n"));
+        break;
       }
-      DDS::ConditionSeq conditions;
-      ret = ws->wait(conditions, timeout);
-      if (ret != DDS::RETCODE_OK && ret != DDS::RETCODE_TIMEOUT) {
-        if (log_level >= LogLevel::Error) {
-          ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: main(): wait failed!\n"));
+
+      DDS::PublicationBuiltinTopicDataSeq pub_bit_seq;
+      DDS::SampleInfoSeq info;
+      pub_reader_i->read(pub_bit_seq, info, DDS::LENGTH_UNLIMITED,
+        DDS::NOT_READ_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ALIVE_INSTANCE_STATE);
+      DynamicTopic* topic = 0;
```

Comment:
This may be something I just can't "see" very well from the diff... but does `topic` need to have scope/lifetime outside of the loop?